File: Seed/Scripts/secret_scan.py
#!/usr/bin/env python3
"""Small local secret scanner for Business AI Starter Kit workspaces."""

from __future__ import annotations

import argparse
import os
import re
import subprocess
import sys
from pathlib import Path


PATTERNS = [
    ("OpenAI API key", re.compile(r"\bsk-[A-Za-z0-9_\-]{20,}\b")),
    ("GitHub token", re.compile(r"\bgh[pousr]_[A-Za-z0-9_]{20,}\b")),
    ("Slack token", re.compile(r"\bxox[baprs]-[A-Za-z0-9\-]{20,}\b")),
    ("Stripe secret key", re.compile(r"\bsk_(live|test)_[A-Za-z0-9]{20,}\b")),
    ("AWS access key", re.compile(r"\bAKIA[0-9A-Z]{16}\b")),
    ("Private key block", re.compile(r"-----BEGIN (RSA |EC |OPENSSH |DSA )?PRIVATE KEY-----")),
    ("Generic secret assignment", re.compile(r"(?i)\b(api[_-]?key|secret|token|password)\b\s*[:=]\s*['\"]?([A-Za-z0-9_./+=\-]{24,})")),
]

SAFE_PLACEHOLDER = re.compile(r"<SECRET:[A-Z0-9_]+>")
SKIP_DIRS = {
    ".git",
    ".business-ai-kit/source",
    "node_modules",
    "__pycache__",
    ".venv",
    "venv",
    ".next",
    "dist",
    "build",
}
SKIP_SUFFIXES = {
    ".png",
    ".jpg",
    ".jpeg",
    ".gif",
    ".webp",
    ".pdf",
    ".zip",
    ".tar",
    ".gz",
    ".sqlite",
    ".db",
}
SAFE_FILENAMES = {".env.example"}


def run_git(args: list[str]) -> list[str]:
    result = subprocess.run(["git", *args], check=False, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if result.returncode != 0:
        return []
    return [line.strip() for line in result.stdout.splitlines() if line.strip()]


def staged_files() -> list[Path]:
    files = run_git(["diff", "--cached", "--name-only", "--diff-filter=ACMR"])
    return [Path(path) for path in files]


def should_skip(path: Path) -> bool:
    path_text = path.as_posix()
    if path.name in SAFE_FILENAMES:
        return True
    if path.suffix.lower() in SKIP_SUFFIXES:
        return True
    return any(path_text == item or path_text.startswith(f"{item}/") for item in SKIP_DIRS)


def candidate_files(paths: list[str], staged: bool) -> list[Path]:
    if staged:
        return [path for path in staged_files() if path.exists() and not should_skip(path)]

    candidates: list[Path] = []
    roots = [Path(path) for path in paths] if paths else [Path(".")]
    for root in roots:
        if root.is_file() and not should_skip(root):
            candidates.append(root)
        elif root.is_dir():
            for dirpath, dirnames, filenames in os.walk(root):
                dirnames[:] = [name for name in dirnames if not should_skip(Path(dirpath, name))]
                for filename in filenames:
                    path = Path(dirpath, filename)
                    if not should_skip(path):
                        candidates.append(path)
    return candidates


def scan_file(path: Path) -> list[tuple[int, str]]:
    try:
        text = path.read_text(encoding="utf-8", errors="ignore")
    except OSError:
        return []

    findings: list[tuple[int, str]] = []
    for line_no, line in enumerate(text.splitlines(), start=1):
        if SAFE_PLACEHOLDER.search(line):
            continue
        for label, pattern in PATTERNS:
            if pattern.search(line):
                findings.append((line_no, label))
                break
    return findings


def main() -> int:
    parser = argparse.ArgumentParser(description="Scan files for likely secrets.")
    parser.add_argument("paths", nargs="*", help="Files or directories to scan. Defaults to current directory.")
    parser.add_argument("--staged", action="store_true", help="Scan staged Git files only.")
    args = parser.parse_args()

    files = candidate_files(args.paths, args.staged)
    all_findings: list[tuple[Path, int, str]] = []
    for path in files:
        for line_no, label in scan_file(path):
            all_findings.append((path, line_no, label))

    if all_findings:
        print("Secret scan failed. Review these files before committing:")
        for path, line_no, label in all_findings:
            print(f"- {path}:{line_no} ({label})")
        print("Replace raw values with <SECRET:NAME> and store values in .env or Doppler.")
        return 1

    print(f"Secret scan passed ({len(files)} file(s) checked).")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: Seed/Scripts/secret_scan.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';

const patterns = [
  ['OpenAI API key', /\bsk-[A-Za-z0-9_-]{20,}\b/],
  ['GitHub token', /\bgh[pousr]_[A-Za-z0-9_]{20,}\b/],
  ['Slack token', /\bxox[baprs]-[A-Za-z0-9-]{20,}\b/],
  ['Stripe secret key', /\bsk_(live|test)_[A-Za-z0-9]{20,}\b/],
  ['AWS access key', /\bAKIA[0-9A-Z]{16}\b/],
  ['Private key block', /-----BEGIN (RSA |EC |OPENSSH |DSA )?PRIVATE KEY-----/],
  ['Generic secret assignment', /\b(api[_-]?key|secret|token|password)\b\s*[:=]\s*['"]?([A-Za-z0-9_./+=-]{24,})/i],
];
const skipDirs = new Set(['.git', 'node_modules', '__pycache__', '.venv', 'venv', '.next', 'dist', 'build']);
const binaryExtensions = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.pdf', '.zip', '.tar', '.gz', '.sqlite', '.db']);

function git(args, cwd = process.cwd(), encoding = 'utf8') {
  const result = spawnSync('git', args, { cwd, encoding, maxBuffer: 32 * 1024 * 1024 });
  if (result.error || result.status !== 0) {
    throw new Error('Git scan input unavailable. Use explicit file paths outside a Git workspace.');
  }
  return result.stdout;
}

function gitPaths(args, root) {
  return git(args, root).split('\0').filter(Boolean);
}

function skip(file) {
  const parts = file.split(path.sep);
  return parts.some((part) => skipDirs.has(part))
    || parts.join('/').includes('.business-ai-kit/source/')
    || binaryExtensions.has(path.extname(file).toLowerCase());
}

function walk(file, files) {
  if (skip(file)) return;
  const stat = fs.lstatSync(file);
  if (stat.isSymbolicLink()) return;
  if (stat.isFile()) files.push(file);
  else if (stat.isDirectory()) {
    for (const name of fs.readdirSync(file)) walk(path.join(file, name), files);
  }
}

function inputs(args) {
  const staged = args.includes('--staged');
  const all = args.includes('--all');
  if (staged && all) throw new Error('Choose --staged or --all, not both.');
  const paths = args.filter((arg) => !['--', '--staged', '--all'].includes(arg));
  if (paths.some((arg) => arg.startsWith('--'))) throw new Error('Unknown scan option.');
  // --staged takes precedence over paths from a package script.
  if (paths.length && !staged && !all) {
    const files = [];
    for (const file of paths) walk(file, files);
    return { root: process.cwd(), files, staged: false };
  }
  const root = git(['rev-parse', '--show-toplevel']).trim();
  const stagedPaths = () => gitPaths(['diff', '--cached', '--name-only', '-z', '--diff-filter=ACMRT'], root);
  const files = staged ? stagedPaths() : all
    ? gitPaths(['ls-files', '-z', '--cached', '--others', '--exclude-standard'], root)
    : [
      ...stagedPaths(),
      ...gitPaths(['diff', '--name-only', '-z', '--diff-filter=ACMRT'], root),
      ...gitPaths(['ls-files', '-z', '--others', '--exclude-standard'], root),
    ];
  return { root, files, staged };
}

function read(file, root, staged) {
  if (staged) return git(['show', `:${file}`], root, null);
  const absolute = path.resolve(root, file);
  // Deleted files and symlinks contain no working-tree text to inspect.
  if (!fs.existsSync(absolute) || fs.lstatSync(absolute).isSymbolicLink()) return null;
  return fs.readFileSync(absolute);
}

function main() {
  const args = process.argv.slice(2);
  if (args.includes('--help') || args.includes('-h')) {
    console.log('Usage: node Scripts/secret_scan.mjs [--staged | --all | paths...]\nDefault: changed and new non-ignored Git files. --staged scans index contents; --all audits all non-ignored Git files.');
    return 0;
  }
  try {
    const { root, files, staged } = inputs(args);
    let checked = 0;
    let found = false;
    for (const file of new Set(files)) {
      if (skip(file)) continue;
      const data = read(file, root, staged);
      if (!data || data.includes(0)) continue;
      checked += 1;
      data.toString('utf8').split(/\r?\n/).forEach((line, index) => {
        // Ignore the placeholder itself, never the rest of a line or example file.
        const content = line.replace(/<SECRET:[A-Z0-9_]+>/g, '');
        const match = patterns.find(([, pattern]) => pattern.test(content));
        if (match) {
          found = true;
          console.log(`- ${JSON.stringify(file)}:${index + 1} (${match[0]})`);
        }
      });
    }
    if (found) {
      console.log('Secret scan failed. Remove raw credentials from these files before sharing or committing.');
      return 1;
    }
    console.log(checked ? `Secret scan passed (${checked} file(s) checked).` : 'Secret scan skipped: no eligible changed files.');
    return 0;
  } catch (error) {
    // Report no file contents or Git output, which could contain credentials.
    console.error(`Secret scan could not complete: ${error.code || error.message}`);
    return 1;
  }
}

process.exit(main());
